README
NAME

    LWPx::UserAgent::Cached - Subclass of LWP::UserAgent that caches HTTP
    GET requests

VERSION

    version 0.007

SYNOPSIS

        use LWPx::UserAgent::Cached;
        use CHI;
    
        my $ua = LWPx::UserAgent::Cached->new(
            cache => CHI->new(
                driver => 'File', root_dir => '/tmp/cache', expires_in => '1d',
            ),
        );
        $ua->get('http://www.perl.org/');

DESCRIPTION

    This module borrows the caching logic from WWW::Mechanize::Cached but
    without inheriting from WWW::Mechanize; instead it is just a direct
    subclass of LWP::UserAgent.

 HTTP/1.1 cache operation

    Full HTTP/1.1 cache compliance is a work in progress. As of version
    0.006 we have limited support for HTTP/1.1 ETag/If-None-Match headers,
    as well as no-cache and no-store Cache-Control directives (both on
    request and response) and the Pragma: no-cache request header.

ATTRIBUTES

 cache

    Settable at construction, defaults to using CHI::Driver::RawMemory with
    an instance-specific hash datastore and a namespace with the current
    package name. You can use your own caching object here as long as it
    has get and set methods.

 is_cached

    Read-only accessor that indicates if the current request is cached or
    not.

 cache_undef_content_length

    Settable at construction or anytime thereafter, indicates whether we
    should cache content even if the HTTP Content-Length header is missing
    or undefined. Defaults to false.

 cache_zero_content_length

    Settable at construction or anytime thereafter, indicates whether we
    should cache content even if the HTTP Content-Length header is zero.
    Defaults to false.

 cache_mismatch_content_length

    Settable at construction or anytime thereafter, indicates whether we
    should cache content even if the length of the data does not match the
    HTTP Content-Length header. Defaults to true.

 ref_in_cache_key

    Settable at construction or anytime thereafter, indicates whether we
    should store the HTTP referrer in the cache key. Defaults to false.

 positive_cache

    Settable at construction or anytime thereafter, indicates whether we
    should only cache positive responses (HTTP response codes from 200 to
    300 inclusive) or cache everything. Defaults to true.

SEE ALSO

    LWP::UserAgent

      Parent of this class.

    WWW::Mechanize::Cached

      Inspiration for this class.

HANDLERS

    This module works by adding request_send, response_done and
    response_header handlers that run on successful HTTP GET requests. If
    you need to modify or remove these handlers you may use
    LWP::UserAgent's handler-related methods.

SUPPORT

 Perldoc

    You can find documentation for this module with the perldoc command.

      perldoc LWPx::UserAgent::Cached

 Websites

    The following websites have more information about this module, and may
    be of help to you. As always, in addition to those websites please use
    your favorite search engine to discover more resources.

      * MetaCPAN

      A modern, open-source CPAN search engine, useful to view POD in HTML
      format.

      http://metacpan.org/release/LWPx-UserAgent-Cached

      * Search CPAN

      The default CPAN search engine, useful to view POD in HTML format.

      http://search.cpan.org/dist/LWPx-UserAgent-Cached

      * AnnoCPAN

      The AnnoCPAN is a website that allows community annotations of Perl
      module documentation.

      http://annocpan.org/dist/LWPx-UserAgent-Cached

      * CPAN Ratings

      The CPAN Ratings is a website that allows community ratings and
      reviews of Perl modules.

      http://cpanratings.perl.org/d/LWPx-UserAgent-Cached

      * CPAN Forum

      The CPAN Forum is a web forum for discussing Perl modules.

      http://cpanforum.com/dist/LWPx-UserAgent-Cached

      * CPANTS

      The CPANTS is a website that analyzes the Kwalitee ( code metrics )
      of a distribution.

      http://cpants.cpanauthors.org/dist/LWPx-UserAgent-Cached

      * CPAN Testers

      The CPAN Testers is a network of smokers who run automated tests on
      uploaded CPAN distributions.

      http://www.cpantesters.org/distro/L/LWPx-UserAgent-Cached

      * CPAN Testers Matrix

      The CPAN Testers Matrix is a website that provides a visual overview
      of the test results for a distribution on various Perls/platforms.

      http://matrix.cpantesters.org/?dist=LWPx-UserAgent-Cached

      * CPAN Testers Dependencies

      The CPAN Testers Dependencies is a website that shows a chart of the
      test results of all dependencies for a distribution.

      http://deps.cpantesters.org/?module=LWPx::UserAgent::Cached

 Bugs / Feature Requests

    Please report any bugs or feature requests through the web interface at
    https://github.com/mjgardner/LWPx-UserAgent-Cached/issues. You will be
    automatically notified of any progress on the request by the system.

 Source Code

    The code is open to the world, and available for you to hack on. Please
    feel free to browse it and play with it, or whatever. If you want to
    contribute patches, please send me a diff or prod me to pull from your
    repository :)

    https://github.com/mjgardner/LWPx-UserAgent-Cached

      git clone git://github.com/mjgardner/LWPx-UserAgent-Cached.git

AUTHOR

    Mark Gardner <mjgardner@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2015 by ZipRecruiter.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for perl distribution LWPx-UserAgent-Role-Cached

{{$NEXT}}

0.011     2020-10-18 20:58:35-05:00 America/Chicago
    - remove AnnoCPAN links from docs
0.010     2018-12-05 12:23:13-06:00 America/Chicago
    - add ignore_headers option to allow for aggressive caching that ignores
      response headers to the contrary
    - no longer strip content headers when caching; this was causing problems
      with some apps that assumed that Content-type would continue to exist
0.009     2018-12-03 16:24:31-06:00 America/Chicago
    - remove link to CPAN forum from docs as it's apparently been discontinued,
      and other doc changes as created by updates to
      Pod::Weaver::Section::Support
0.008     2016-10-27 00:29:55-05:00 America/Chicago
    - more correctly handle HTTP Client-Transfer-Encoding: chunked
      headers
    - remove Modern::Perl from prerequisites
    - documentation tweaks (automatically generating list of
      prerequisites, listing other Perl modules in monospace text)
0.007     2015-03-04 17:06:36-05:00 America/New_York
    - documentation and metadata fixes, no functional changes
0.006     2015-03-03 13:34:06-05:00 America/New_York
    - use CHI's built-in serializer support by default (with Sereal) instead
      of serializing in the UserAgent
    - fix and fully test ETag support
    - add support for no-cache and no-store Cache-Control request and response
      headers, and Pragma: no-cache request header
0.005     2015-02-17 22:59:02-05:00 America/New_York (TRIAL RELEASE)
    - add preliminary ETag cache control support
0.004     2015-02-12 22:55:49-05:00 America/New_York
    - require newer version of Moo's Sub::Quote to get the qsub function
0.003     2015-02-11 22:15:47-05:00 America/New_York
    - more micro-optimizations with Sereal replacing Storable
0.002     2015-02-10 21:40:24-05:00 America/New_York (TRIAL RELEASE)
    - require a version of Moo that allows undef defaults in attributes
    - use faster Type::Tiny rather than MooX::Types::MooseLike
0.001     2015-02-10 18:22:57-05:00 America/New_York (TRIAL RELEASE)
    - Initial release


t/TestCache.pm
package TestCache;

use Sereal qw(encode_sereal decode_sereal);

sub new {
    my $class = shift;

    my $self = {};

    bless $self, $class;

    return $self;
}

sub set {
    my $self  = shift;
    my $key   = shift;
    my $value = shift;    # an HTTP::Response

    my $res = decode_sereal($value);
    $res->content("DUMMY");
    $self->{$key} = encode_sereal($res);
}

sub get {
    my $self = shift;
    my $key  = shift;

    return $self->{$key};
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.032.
use strict;
use warnings;

use 5.012;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Subclass of LWP::UserAgent that caches HTTP GET requests",
  "AUTHOR" => "Mark Gardner <mjgardner\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "LWPx-UserAgent-Cached",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.012",
  "NAME" => "LWPx::UserAgent::Cached",
  "PREREQ_PM" => {
    "CHI" => 0,
    "Data::Serializer::Sereal" => 0,
    "HTTP::Status" => 0,
    "LWP::UserAgent" => 0,
    "List::Util" => "1.33",
    "Modern::Perl" => 0,
    "Moo" => "1.004005",
    "Types::Standard" => 0,
    "namespace::clean" => 0,
    "utf8" => 0
  },
  "TEST_REQUIRES" => {
    "Carp" => 0,
    "Const::Fast" => 0,
    "Digest::SHA" => 0,
    "English" => 0,
    "File::Spec" => 0,
    "Find::Lib" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Path::Class" => 0,
    "Scalar::Util" => 0,
    "Test::Fake::HTTPD" => 0,
    "Test::More" => "0.94",
    "Test::Requires" => 0,
    "constant" => 0,
    "lib" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.007",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CHI" => 0,
  "Carp" => 0,
  "Const::Fast" => 0,
  "Data::Serializer::Sereal" => 0,
  "Digest::SHA" => 0,
  "English" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Find::Lib" => 0,
  "HTTP::Status" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "LWP::UserAgent" => 0,
  "List::Util" => "1.33",
  "Modern::Perl" => 0,
  "Moo" => "1.004005",
  "Path::Class" => 0,
  "Scalar::Util" => 0,
  "Test::Fake::HTTPD" => 0,
  "Test::More" => "0.94",
  "Test::Requires" => 0,
  "Types::Standard" => 0,
  "constant" => 0,
  "lib" => 0,
  "namespace::clean" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/LWPx/UserAgent/Cached.pm
package LWPx::UserAgent::Cached;

# ABSTRACT: Subclass of LWP::UserAgent that caches HTTP GET requests

use strict;
use warnings;
use utf8;
our $VERSION = '0.012';

## no critic (Bangs::ProhibitCommentedOutCode)

=head1 SYNOPSIS

    use LWPx::UserAgent::Cached;
    use CHI;

    my $ua = LWPx::UserAgent::Cached->new(
        cache => CHI->new(
            driver => 'File', root_dir => '/tmp/cache', expires_in => '1d',
        ),
    );
    $ua->get('http://www.perl.org/');

=head1 DESCRIPTION

This module borrows the caching logic from
L<C<WWW::Mechanize::Cached>|WWW::Mechanize::Cached> but
without inheriting from
L<C<WWW::Mechanize>|WWW::Mechanize>; instead it is just
a direct subclass of
L<C<LWP::UserAgent>|LWP::UserAgent>.

=head2 HTTP/1.1 cache operation

Full HTTP/1.1 cache compliance is a work in progress. As of version 0.006 we
have limited support for HTTP/1.1 C<ETag>/C<If-None-Match> headers, as well as
C<no-cache> and C<no-store> C<Cache-Control> directives (both on request and
response) and the C<Pragma: no-cache> request header.

=head1 SEE ALSO

=over

=item L<C<LWP::UserAgent>|LWP::UserAgent>

Parent of this class.

=item L<C<WWW::Mechanize::Cached>|WWW::Mechanize::Cached>

Inspiration for this class.

=back

=cut

use CHI;
use HTTP::Status qw(HTTP_OK HTTP_MOVED_PERMANENTLY HTTP_NOT_MODIFIED);
use List::Util 1.33 'any';
use Moo 1.004005;
use Types::Standard qw(Bool HasMethods HashRef InstanceOf Maybe);
use namespace::clean;
extends 'LWP::UserAgent';

=attr cache

Settable at construction, defaults to using
L<C<CHI::Driver::RawMemory>|CHI::Driver::RawMemory> with
an instance-specific hash datastore and a namespace with the current
package name. You can use your own caching object here as long as it has
C<get> and C<set> methods.

=cut

has cache => (
    is      => 'lazy',
    isa     => HasMethods [qw(get set)],
    default => sub {
        CHI->new(
            serializer => 'Sereal',
            driver     => 'RawMemory',
            datastore  => $_[0]->_cache_datastore,
            namespace  => __PACKAGE__,
        );
    },
);
has _cache_datastore =>
    ( is => 'lazy', isa => HashRef, default => sub { {} } );

=attr is_cached

Read-only accessor that indicates if the current request is cached or not.

=cut

has is_cached =>
    ( is => 'rwp', isa => Maybe [Bool], init_arg => undef, default => undef );

=attr cache_undef_content_length

Settable at construction or anytime thereafter, indicates whether we should
cache content even if the HTTP C<Content-Length> header is missing or
undefined. Defaults to false.

=cut

has cache_undef_content_length => ( is => 'rw', isa => Bool, default => 0 );

=attr cache_zero_content_length

Settable at construction or anytime thereafter, indicates whether we should
cache content even if the HTTP C<Content-Length> header is zero. Defaults to
false.

=cut

has cache_zero_content_length => ( is => 'rw', isa => Bool, default => 0 );

=attr cache_mismatch_content_length

Settable at construction or anytime thereafter, indicates whether we should
cache content even if the length of the data does not match the HTTP
C<Content-Length> header. Defaults to true.

=cut

has cache_mismatch_content_length =>
    ( is => 'rw', isa => Bool, default => 1 );

=attr ref_in_cache_key

Settable at construction or anytime thereafter, indicates whether we should
store the HTTP referrer in the cache key. Defaults to false.

=cut

has ref_in_cache_key => ( is => 'rw', isa => Bool, default => 0 );

=attr positive_cache

Settable at construction or anytime thereafter, indicates whether we should
only cache positive responses (HTTP response codes from C<200> to C<300>
inclusive) or cache everything. Defaults to true.

=cut

has positive_cache => ( is => 'rw', isa => Bool, default => 1 );

=attr ignore_headers

Settable at construction or anytime thereafter, indicates whether we should
ignore C<Cache-Control: no-cache>, C<Cache-Control: no-store>, and
C<Pragma: no-cache> HTTP headers when deciding whether to cache a response.
Defaults to false.

B<Important note:> This option is potentially dangerous, as it ignores the
explicit instructions from the server and thus can lead to returning stale
content.

=cut

has ignore_headers => ( is => 'rw', isa => Bool, default => 0 );

=head1 HANDLERS

This module works by adding C<request_send>, C<response_done> and
C<response_header> L<handlers|LWP::UserAgent/Handlers>
that run on successful HTTP C<GET> requests.
If you need to modify or remove these handlers you may use LWP::UserAgent's
L<handler-related methods|LWP::UserAgent/Handlers>.

=for Pod::Coverage BUILD

=cut

sub BUILD {
    my $self = shift;

    $self->add_handler( request_send => \&_get_cache, ( m_method => 'GET' ) );
    $self->add_handler(
        response_done => \&_set_cache,
        ( m_method => 'GET', m_code => 2 ),
    );
    $self->add_handler(
        response_header => \&_get_not_modified,
        ( m_method => 'GET', m_code => HTTP_NOT_MODIFIED ),
    );

    return;
}

# load from cache on each GET request
sub _get_cache {
    my ( $request, $self ) = @_;
    $self->_set_is_cached(0);

    my $clone = $request->clone;
    if ( not $self->ref_in_cache_key ) { $clone->header( Referer => undef ) }
    return if $self->_no_cache_header_directives($request);

    return if not my $response = $self->cache->get( $clone->as_string );
    return
        if $response->code < HTTP_OK
        or $response->code > HTTP_MOVED_PERMANENTLY;

    if ( $response->header('etag') ) {
        $clone->header( if_none_match => $response->header('etag') );
        $response = $self->request($clone);
    }
    return if $self->_no_cache_header_directives($response);

    $self->_set_is_cached(1);
    return $response;
}

sub _get_not_modified {
    my ( $response, $self ) = @_;
    $self->_set_is_cached(0);

    my $request = $response->request->clone;
    $request->remove_header(qw(if_modified_since if_none_match));

    my $cached_response = $self->cache->get( $request->as_string );
    $response->content( $cached_response->decoded_content );

    $self->_set_is_cached(1);
    return;
}

# save to cache after successful GET
sub _set_cache {
    my ( $response, $self ) = @_;
    return if not $response;

    if (not($response->header('client-transfer-encoding')
            and any { 'chunked' eq $_ }
            $response->header('client-transfer-encoding')
        )
        )
    {
        for ( $response->header('size') ) {
            return
                if not defined and $self->cache_undef_content_length;
            return
                if 0 == $_
                and not $self->cache_zero_content_length;
            return
                if $_ != length $response->content
                and not $self->cache_mismatch_content_length;
        }
    }

    for my $message ( $response, $response->request ) {
        return if $self->_no_cache_header_directives($message);
    }

    $response->decode;
    $self->cache->set( $response->request->as_string => $response );
    return;
}

sub _no_cache_header_directives {
    my ( $self, $message ) = @_;
    return if $self->ignore_headers;

    for my $header_name (qw(pragma cache_control)) {
        if ( my @directives = $message->header($header_name) ) {
            return 1 if any {/\A no- (?: cache | store ) /xms} @directives;
        }
    }
    return;
}

=for Pod::Coverage FOREIGNBUILDARGS

=cut

## no critic (Subroutines::RequireArgUnpacking)
sub FOREIGNBUILDARGS {
    shift;
    return 'HASH' eq ref $_[0] ? %{ $_[0] } : @_;
}

1;
